js/runner.go

```diff
@@ -113,6 +116,97 @@ func newFromBundle(logger *logrus.Logger, b *Bundle) (*Runner, error) {
 	return r, err
 }
 
+type IpPool struct {
+        ip       net.IP
+        hostStart uint64
+        netStart  uint64
+        hostRangeNum  uint64
+        netRangeNum   uint64
+}
+
+func GetIpPool(s string) *IpPool {
+        for i := 0; i < len(s); i++ {
+                switch s[i] {
+                case '-':
+                        return ipPoolFromRange(s)
+                case '/':
+                        return ipPoolFromCidr(s)
+                }
+        }
+        return nil
+}
+
+func ipPoolFromRange(s string) *IpPool {
+        ss := strings.SplitN(s, "-", 2)
+        ip0, ip1 := net.ParseIP(ss[0]), net.ParseIP(ss[1]) // len(ParseIP())==16
+        if ip0 == nil || ip1 == nil {
+                fmt.Println("Wrong IP range format, exit.", s)
+                return nil
+        }
+        n0 := binary.BigEndian.Uint64(ip0[:8])
+        h0 := binary.BigEndian.Uint64(ip0[8:])
+        n1 := binary.BigEndian.Uint64(ip1[:8])
+        h1 := binary.BigEndian.Uint64(ip1[8:])
+        if (n0 > n1) || (h0 > h1) {
+                fmt.Println("Negative IP range, exit.", s)
+                return nil
+        }
+        return &IpPool{ip0, h0, n0, h1 - h0 + 1, n1 - n0 + 1}
+}
+
+func ipPoolFromCidr(s string) *IpPool {
+        ip0, pnet, err := net.ParseCIDR(s)
+        if err != nil {
+                fmt.Println("ParseCIDR() failed: ", s)
+                return nil
+        }
+        // ip0 (range start) >= ip1 (cidr start)
+        ip1 := pnet.IP.To16()
+        n0 := binary.BigEndian.Uint64(ip0[:8])
+        h0 := binary.BigEndian.Uint64(ip0[8:])
+        n1 := binary.BigEndian.Uint64(ip1[:8])
+        h1 := binary.BigEndian.Uint64(ip1[8:])
+        ones, bits := pnet.Mask.Size()
+        mask := bits - ones
+        hostPos, netPos := h0-h1, n0-n1
+        netRangeNum, hostRangeNum := uint64(0), uint64(1<<mask)-hostPos
+        if mask > 64 {
+                netRangeNum, hostRangeNum = uint64(1<<(mask-64))-netPos, ^uint64(0)-hostPos
+        }
+        return &IpPool{ip0, h0, n0, hostRangeNum, netRangeNum}
+}
+
+// GetRandomIP return a random-by-seed selected IP from an IP pool
+// IPv4 pool is considered as a uniform distribution from start IP to end IP
+// IPv6 poop is considered as two independent uniform distributions: high 64-bit and low 64-bit
+func (pool *IpPool) GetRandomIP(seed int64) net.IP {
+        var hostShift, netShift uint64
+        r := rand.New(rand.NewSource(seed))
+        if pool.hostRangeNum > 0 {
+                hostShift = r.Uint64() % pool.hostRangeNum
+        }
```

Comment:
1. sometimes VUs number is more than IP addrs available so they must share IPs, that is why I use rand() % ip_num.   for 1:1 ordered binding, specify a mode to not using rand() but only using vu_id % ip_num?
2. instead of calculating, unrolling all range IPs to a linear array may require too much memory if the range is very big - especially with CIDR format. 
3. for multiple ranges, how to map VUs to all possible IPs? the later ranges will be less possibility to be picked if append to array list.

Reply:
@MStoykov Will VUs be able to change their local IP after initialisation? looks like NewVU() is called only in scale up stage

Reply:
it can match requirements like:
1. multiple IPs: ip1,ip2,ip3
2. multiple IPs and a range: ip1,ip2,ip3,ip4-ip10|0|6
3. multiple ranges: ip1-ip8,ip10/29